### ISSUE TITLE
XM_CALLCONV

### ISSUE THREAD

**Issue by matt77hias** [Microsoft/DirectXMath#63]
[MSDN](https://msdn.microsoft.com/en-us/library/windows/desktop/ee418728(v=vs.85).aspx):

> In addition to the type aliases, you must also use the `XM_CALLCONV` annotation to make sure the function uses the appropriate calling convention (`__fastcall` versus `__vectorcall`) based on your compiler and architecture. Because of limitations with `__vectorcall`, we recommend that you not use `XM_CALLCONV` for C++ constructors.

[MSDN](https://msdn.microsoft.com/en-us/library/dn375768.aspx):

> The `__vectorcall` calling convention specifies that **arguments to functions** are to be passed in registers, when possible. `__vectorcall` uses more registers for arguments than `__fastcall` or the default x64 calling convention use. The `__vectorcall` calling convention is only supported in native code on x86 and x64 processors that include Streaming SIMD Extensions 2 (SSE2) and above. Use `__vectorcall` to speed functions that pass several floating-point or SIMD vector arguments and perform operations that take advantage of the arguments loaded in registers.

Given these two documentation snippets, I wonder why `XMLoadX` (e.g. `XMLoadFloat2`, etc.) use `XM_CALLCONV`, since no `XMVECTOR` or `XMMATRIX` (or some type aliases) are passed as arguments for these functions?

**Comment by walbourn** [comment]
The load functions are annotated with ``XM_CALLCONV`` to take advantage of the ``__vectorcall`` register semantics for return values.  This [blog post](https://blogs.msdn.microsoft.com/vcblog/2013/07/11/introducing-vector-calling-convention/) talks a bit about the convention including this comment:

> The return value of type HVA/HFA is returned via XMM0(/YMM0): XMM3(/YMM3), one register per element.

The [MSDN __vectorcall](https://msdn.microsoft.com/en-us/library/dn375768.aspx) also states:

> Results of __vectorcall functions are returned by value in registers when possible. Results of integer type, including structs or unions of 8 bytes or less, are returned by value in RAX. Vector type results are returned by value in XMM0 or YMM0, depending on size. HVA results have each data element returned by value in registers XMM0:XMM3 or YMM0:YMM3, depending on element size. Result types that don't fit in the corresponding registers are returned by reference to memory allocated by the caller.

**TL;DR:** Normally in the x64 ``__fastcall`` the ``XMM0L`` is only used to return ``float`` or ``double`` values, so vector return types are returned on the stack with ``RAX`` pointing to that address. With ``__fastcall`` it can return ``XMVECTOR`` in ``XMM0`` and ``XMMATRIX`` in ``XMM0:XMM3``.

**Comment by matt77hias** [comment]
Thanks for the clarification.
